File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void solve(string &digits,string map[],string &output,vector<string>&ans,int ind){
        if(ind>=digits.size()){
            ans.push_back(output);
            return;
        }
        int num=digits[ind]-'0';
        string str=map[num];
        if(str.empty())solve(digits,map,output,ans,ind+1);
        for(int i=0;i<str.size();i++){
            output.push_back(str[i]);
            solve(digits,map,output,ans,ind+1);
            output.pop_back();
        }
    }

    vector<string> letterCombinations(string digits) {
        string map[10]={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        string output="";
        vector<string>ans;
        solve(digits,map,output,ans,0);
        return ans;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public void solve(String digits,String[] map,StringBuilder output,List<String>ans,int ind){
        if(ind>=digits.length()){
            ans.add(output.toString());
            return;
        }
        int num=digits.charAt(ind)-'0';
        if(map[num].isEmpty())solve(digits,map,output,ans,ind+1);
        String str=map[num];
        for(int i=0;i<str.length();i++){
            output.append(str.charAt(i));
            solve(digits,map,output,ans,ind+1);
            output.deleteCharAt(output.length()-1);
        }
    }

    public List<String> letterCombinations(String digits) {
        List<String>ans=new ArrayList<>();
        StringBuilder output=new StringBuilder();
        String[] map={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        solve(digits,map,output,ans,0);
        return ans;
    }
}